### PR TITLE
MBS-13727, MBS-13731: Event edit form fixes

### PR DESCRIPTION
**Pull request by mwiencek**
# Problem

MBS-13727: Unable to submit corrected event setlist
MBS-13731: Cannot add new events from "Add event" popup

# Solution

See commits for details.

The diff is best viewed without whitespace changes.

# Testing

Manual testing + added Selenium test cases.

# Further action

I suggest that we hotfix this to production.

### CODE REVIEW

**Comment by mwiencek** [comment]
Found some very fun behavior, if you do:
```JS
export default component Component() {
  return null;
}; // <-- semi-colon here?
```
It gives two contradictory warnings:
>  3:2  warning  Missing semicolon      @stylistic/js/semi
>  3:2  warning  Unnecessary semicolon  @stylistic/js/no-extra-semi

**Review by reosarevok — COMMENTED** [review]
FWIW, I don't think the large `Add event` header was supposed to be shown within dialogs either, although that seems more annoying to disable.

The event dialog cannot be closed with `Esc` like others can - I don't know if that's a new issue introduced here though or it was introduced with my original PR.

The main issues this is trying to fix do seem to be fixed though, so if you feel like hotfixing this first and then looking into the other stuff more calmly, that could work.

**Comment by mwiencek** [comment]
Thanks, those seem less urgent, so I'll start with hotfixing this for now.

There's also MBS-13735 which I haven't been able to reproduce locally on master, though I have a feeling it has something to do with root/static/scripts/edit/validation.js still running on the page--we have two separate code paths trying to control the disabled state of the submit button at the same time.

File: lib/MusicBrainz/Server/Controller/Role/Create.pm [hunk 1]
```diff
@@ -83,7 +83,15 @@ role {
 
                 return unless $args{within_dialog};
 
-                $c->stash( dialog_result => $c->json->encode(JSONSerializer->serialize_internal($c, $entity)) );
+                $c->stash(
+                    current_view => 'Node',
+                    component_path => 'forms/DialogResult',
```
Comment (yvanzo) [inline lib/MusicBrainz/Server/Controller/Role/Create.pm:88]: Where is that called? For entity creation page only?

File: root/forms/DialogResult.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+ * This file is part of MusicBrainz, the open internet music database,
+ * and is licensed under the GPL version 2, or (at your option) any
+ * later version: http://www.gnu.org/licenses/gpl-2.0.txt
+ */
+
+import escapeClosingTags from '../utility/escapeClosingTags.js';
+
+component DialogResult(
+  result: string,
+) {
```
Comment (yvanzo) [inline root/forms/DialogResult.js:14]: After the message of your first commit:

> One important utility of root/forms/dialog.tt which I did port to React is in
how it returns the created entity JSON in a global variable on a blank page
after the form is submitted

* What is the name of this _utility_?
* What is the name of this _global variable_?
* Should `root/forms/dialog.tt` have a comment as below?
  ```perl
  # Converted to React at root/forms/DialogResult.js`
  ```

File: root/forms/DialogResult.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+  const script = (
+    'window.dialogResult = ' +
+    escapeClosingTags(result) +
+    ';'
+  );
+  return (
+    <html>
+      <head>
+        <script dangerouslySetInnerHTML={{__html: script}} />
```
Comment (yvanzo) [inline root/forms/DialogResult.js:23]: Can you please remind (in commit message) what Perl equivalent does it replace? And the origin(s) of values that `result` (used to build `script`) is going to take?

File: root/static/scripts/event/components/EventEditForm.js [hunk 4]
```diff
@@ -115,6 +117,22 @@ function reducer(state: StateT, action: ActionT): StateT {
...
+      fieldCtx.update('setlist', (setlistFieldCtx) => {
+        setlistFieldCtx.set('value', action.setlist);
+        if (isValidSetlist(action.setlist)) {
+          setlistFieldCtx.set('has_errors', false);
+          setlistFieldCtx.set('errors', []);
+        } else {
+          setlistFieldCtx.set('has_errors', true);
+          setlistFieldCtx.set('errors', [
+            l(`Please ensure all lines start with @, * or #,
+                followed by a space.`),
```
Comment (yvanzo) [inline root/static/scripts/event/components/EventEditForm.js:130]: Untested minor formatting suggestion:

```suggestion
            l(`Please ensure all lines start with
               <code>@</code>, <code>*</code> or <code>#</code>,
               followed by a space.`),
```

**Review by yvanzo — COMMENTED** [review]
Thank you for fixing these regressions and adding tests.

The first edit is hard to follow, pretty important when using `dangerouslySetInnerHTML`. See comments. Should it be split into two parts?

Left a few other minor comments.

File: lib/MusicBrainz/Server/Controller/Role/Create.pm (hunk 1 above)
Comment (mwiencek) [reply lib/MusicBrainz/Server/Controller/Role/Create.pm:88]: Correct, but more specifically, for entity creation forms *inside a dialog* after the form is submitted from within the dialog.

File: root/forms/DialogResult.js (hunk 2 above)
Comment (mwiencek) [reply root/forms/DialogResult.js:14]: 1. By "utility" I actually meant "use" (noun), will fix the wording. So the use is just what's described in the sentence.
2. The name of the global variable is `dialogResult`.
3. Yes, I'll add that comment and clarify the commit message too.

File: root/static/scripts/event/components/EventEditForm.js (hunk 4 above)
Comment (mwiencek) [reply root/static/scripts/event/components/EventEditForm.js:130]: I planned to commit that to beta separately (instead of part of this PR), and it seems to work: https://gist.github.com/mwiencek/f29c6bb81b2d3425d5b392b279e1afe2

However, after testing it I couldn't notice much difference visually, so not 100% sure if it's worth it. But I can commit that if you'd like.

File: root/forms/DialogResult.js (hunk 3 above)
Comment (mwiencek) [reply root/forms/DialogResult.js:23]: I split out 3104264712ca138e9a0888334d96a3cf3bfb398c into its own commit and tried to improve the message. Hope that looks better.

**Comment by yvanzo** [comment]
Thanks a lot for your detailed replies, for your improvements to commit messages and code, and for having deployed it so quickly. And indeed, the split commits are much easier to follow and `dangerouslySetInnerHTML` is legitimate here.

File: root/static/scripts/event/components/EventEditForm.js (hunk 4 above)
Comment (yvanzo) [reply root/static/scripts/event/components/EventEditForm.js:130]: > Edit: In case it wasn't clear, I was copying the existing string from lib/MusicBrainz/Server/Form/Field/Setlist.pm

Okay, I thought that it was a new string. Thanks for having committed this suggestion separately then.

> However, after testing it I couldn't notice much difference visually, so not 100% sure if it's worth it. But I can commit that if you'd like.

This is mainly because we don’t have an appropriate style for displaying `<code>` tags. Additionally to the sans-serif, it should probably have some background color with border radius, light padding and white-space handling; as in GitHub or GitLab for example.
